In [1]:

import pandas as pd
import nltk
import re
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report

#Carga de datos
train_data = pd.read_csv("../../Data/BasedOnSentiments/train_sentiments.csv")
test_data = pd.read_csv("../../Data/BasedOnSentiments/test_sentiments.csv")
valid_data = pd.read_csv("../../Data/BasedOnSentiments/valid_sentiments.csv")

# --- PREPROCESAMIENTO PARA LEMATIZACIÓN ---
#nltk.download('punkt')
#nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    words = re.findall(r'\b\w+\b', text.lower())
    lemmatized_words = [lemmatizer.lemmatize(word) for word in words]
    return " ".join(lemmatized_words)

# Datos
X_train_raw = train_data['Text'].values
X_test_raw = test_data['Text'].values
y_train = train_data['Sentiment'].values
y_test = test_data['Sentiment'].values

# Datos lematizados
X_train_lem = [preprocess_text(text) for text in X_train_raw]
X_test_lem = [preprocess_text(text) for text in X_test_raw]

In [2]:
# --- DIRECTORIO DE SALIDA DE LOS PLOTS ---
output_dir = "../../Plots/Experiment1/Sentiments/Experiment/"


import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix

# Función para graficar matriz de confusión
def plot_confusion_matrix(y_true, y_pred, labels, title):
    cm = confusion_matrix(y_true, y_pred, labels=list(range(len(labels))))

    plt.figure(figsize=(10, 8))
    sns.heatmap(
        cm,
        annot=True,
        fmt="d",
        cmap="Blues", 
        xticklabels=labels,
        yticklabels=labels,
        cbar=True
    )
    plt.title(f'Matriz de Confusión - {title}')
    plt.ylabel('Etiqueta Verdadera')
    plt.xlabel('Etiqueta Predicha')
    plt.xticks(rotation=45)
    plt.yticks(rotation=0)
    plt.tight_layout()
    # Guardar archivo
    filename = title.replace(" ", "_").replace("(", "").replace(")", "").lower() + ".png"
    plt.savefig(output_dir + filename)
    plt.close()

In [3]:
# --- FUNCION GENERAL PARA COMPARAR ---
def train_and_evaluate(model, X_train, y_train, X_test, y_test, label):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(f"\n--- {label} ---")
    classification_report(y_test, y_pred)
    report_dict = classification_report(y_test, y_pred, output_dict=True, zero_division=0)
    f1_macro = report_dict["macro avg"]["f1-score"]
    recall_macro = report_dict["macro avg"]["recall"]
    print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
    print(f"F1 Score (macro avg): {f1_macro:.4f}")
    print(f"Recall Score (macro avg): {recall_macro:.4f}")

    # Mostrar matriz de confusión
    sentiment_labels = ["Positive", "Negative", "Ambiguous", "Neutral"]
    plot_confusion_matrix(y_test, y_pred, sentiment_labels, label)

In [4]:
# --- VECTORIZADORES ---
vectorizer_raw = TfidfVectorizer(max_features=10000)
vectorizer_stopwords = TfidfVectorizer(max_features=10000, stop_words='english')
vectorizer_lowercase = TfidfVectorizer(max_features=10000, lowercase=True)
vectorizer_strip = TfidfVectorizer(max_features=10000, strip_accents='unicode')
vectorizer_stopwords_lowercase = TfidfVectorizer(max_features=10000, stop_words='english', lowercase=True)
vectorizer_strip_lowercase = TfidfVectorizer(max_features=10000, strip_accents='unicode', lowercase=True)
vectorizer_complete = TfidfVectorizer(max_features=10000, stop_words='english', lowercase=True, strip_accents='unicode')

#TRAIN Y TEST CON Y SIN LEMATIZACIÖN (TF-IDF sin hiperparámetros)
X_train_raw_tfidf = vectorizer_raw.fit_transform(X_train_raw)
X_test_raw_tfidf = vectorizer_raw.transform(X_test_raw)

X_train_lem_tfidf = vectorizer_raw.fit_transform(X_train_lem)
X_test_lem_tfidf = vectorizer_raw.transform(X_test_lem)

X_train_lem_tfidf_stopwords = vectorizer_stopwords.fit_transform(X_train_lem)
X_test_lem_tfidf_stopwords = vectorizer_stopwords.transform(X_test_lem)

X_train_lem_tfidf_lowercase = vectorizer_lowercase.fit_transform(X_train_lem)
X_test_lem_tfidf_lowercase = vectorizer_lowercase.transform(X_test_lem)

X_train_lem_tfidf_strip = vectorizer_strip.fit_transform(X_train_lem)
X_test_lem_tfidf_strip = vectorizer_strip.transform(X_test_lem)

X_train_lem_tfidf_stopwords_lowercase = vectorizer_stopwords_lowercase.fit_transform(X_train_lem)
X_test_lem_tfidf_stopwords_lowercase = vectorizer_stopwords_lowercase.transform(X_test_lem)

X_train_lem_tfidf_strip_lowercase = vectorizer_strip_lowercase.fit_transform(X_train_lem)
X_test_lem_tfidf_strip_lowercase = vectorizer_strip_lowercase.transform(X_test_lem)

X_train_lem_tfidf_complete = vectorizer_complete.fit_transform(X_train_lem)
X_test_lem_tfidf_complete = vectorizer_complete.transform(X_test_lem)





In [ ]:
#Modelos de clasificación
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000, random_state=42),
    'SVM (linear)': SVC(kernel='linear', class_weight='balanced'),
    'Decision Tree': DecisionTreeClassifier(random_state=42)
}

SoporteVectorial = SVC(kernel='linear', class_weight='balanced')


In [ ]:
# --- COMPARATIVA FINAL ---
print("\n-- TF-IDF --")
for model_name, model in models.items():
    train_and_evaluate(model, X_train_raw_tfidf, y_train, X_test_raw_tfidf, y_test, f"{model_name} (TF-IDF)")


print("\n-- CON LEMATIZAR --")
for model_name, model in models.items():
    train_and_evaluate(model, X_train_lem_tfidf, y_train, X_test_lem_tfidf, y_test, f"{model_name} (TF-IDF Lematizado)")

print("\nEl experimento con el TF-IDF con lematizar obtuvo mejores resultados\n")
print("Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo\n")
print("Además, el clasificador con mejor desempeño fue el SVM linear, por lo que será el usado en los próximos experimentos")


print("\n--TF-IDF SIN STOPWORDS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_stopwords, y_train, X_test_lem_tfidf_stopwords, y_test, "SVM (linear) (TF-IDF sin stopwords)")

print("\n-- TF-IDF EN MINUSCULAS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_lowercase, y_train, X_test_lem_tfidf_lowercase, y_test, "SVM (linear) (TF-IDF en minusculas)")

print("\n-- TF-IDF SIN ACENTOS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_strip, y_train, X_test_lem_tfidf_strip, y_test, "SVM (linear) (TF-IDF sin acentos)")

print("\n-- TF-IDF SIN STOPWORDS Y EN MINUSCULAS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_stopwords_lowercase, y_train, X_test_lem_tfidf_stopwords_lowercase, y_test, "SVM (linear)  (TF-IDF sin stopwords y en minusculas)")

print("\n-- TF-IDF SIN ACENTOS Y EN MINUSCULAS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_strip_lowercase, y_train, X_test_lem_tfidf_strip_lowercase, y_test, "SVM (linear)  (TF-IDF sin acentos y en minusculas)")

print("\n-- TF-IDF COMPLETO --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_complete, y_train, X_test_lem_tfidf_complete, y_test, "SVM (linear)  (TF-IDF completo)")






-- TF-IDF --

--- Logistic Regression (TF-IDF) ---
Accuracy: 0.6563
F1 Score (macro avg): 0.5866
Recall Score (macro avg): 0.5707

--- SVM (linear) (TF-IDF) ---
Accuracy: 0.6370
F1 Score (macro avg): 0.6027
Recall Score (macro avg): 0.6196

--- Decision Tree (TF-IDF) ---
Accuracy: 0.5327
F1 Score (macro avg): 0.4672
Recall Score (macro avg): 0.4639

-- CON LEMATIZAR --

--- Logistic Regression (TF-IDF Lematizado) ---
Accuracy: 0.6606
F1 Score (macro avg): 0.5928
Recall Score (macro avg): 0.5764

--- SVM (linear) (TF-IDF Lematizado) ---
Accuracy: 0.6374
F1 Score (macro avg): 0.6032
Recall Score (macro avg): 0.6205

--- Decision Tree (TF-IDF Lematizado) ---
Accuracy: 0.5443
F1 Score (macro avg): 0.4751
Recall Score (macro avg): 0.4719

El experimento con el TF-IDF con lematizar obtuvo mejores resultados

Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo

Además, el clasificador con mejor desemp

-- TF-IDF --

--- Logistic Regression (TF-IDF) ---
Accuracy: 0.6563
              precision    recall  f1-score   support

           0       0.78      0.76      0.77      1863
           1       0.67      0.52      0.58      1070
           2       0.57      0.26      0.35       488
           3       0.55      0.75      0.64      1606

    accuracy                           0.66      5027
   macro avg       0.65      0.57      0.59      5027
weighted avg       0.67      0.66      0.65      5027


--- SVM (linear) (TF-IDF) ---
Accuracy: 0.6612
              precision    recall  f1-score   support

           0       0.80      0.75      0.77      1863
           1       0.69      0.52      0.59      1070
           2       0.60      0.25      0.35       488
           3       0.55      0.78      0.65      1606

    accuracy                           0.66      5027
   macro avg       0.66      0.57      0.59      5027
weighted avg       0.68      0.66      0.65      5027


--- Decision Tree (TF-IDF) ---
Accuracy: 0.5327
              precision    recall  f1-score   support

           0       0.68      0.68      0.68      1863
           1       0.45      0.40      0.42      1070
           2       0.28      0.24      0.26       488
           3       0.48      0.54      0.51      1606

    accuracy                           0.53      5027
   macro avg       0.47      0.46      0.47      5027
weighted avg       0.53      0.53      0.53      5027


-- CON LEMATIZAR --

--- Logistic Regression (TF-IDF Lematizado) ---
Accuracy: 0.6606
              precision    recall  f1-score   support

           0       0.78      0.77      0.77      1863
           1       0.68      0.51      0.59      1070
           2       0.58      0.27      0.37       488
           3       0.56      0.75      0.64      1606

    accuracy                           0.66      5027
   macro avg       0.65      0.58      0.59      5027
weighted avg       0.67      0.66      0.65      5027


--- SVM (linear) (TF-IDF Lematizado) ---
Accuracy: 0.6602
              precision    recall  f1-score   support

           0       0.80      0.75      0.77      1863
           1       0.68      0.52      0.59      1070
           2       0.59      0.24      0.34       488
           3       0.55      0.78      0.65      1606

    accuracy                           0.66      5027
   macro avg       0.65      0.57      0.59      5027
weighted avg       0.67      0.66      0.65      5027


--- Decision Tree (TF-IDF Lematizado) ---
Accuracy: 0.5443
              precision    recall  f1-score   support

           0       0.69      0.70      0.69      1863
           1       0.47      0.42      0.44      1070
           2       0.27      0.22      0.25       488
           3       0.49      0.55      0.52      1606

    accuracy                           0.54      5027
   macro avg       0.48      0.47      0.48      5027
weighted avg       0.54      0.54      0.54      5027


El experimento con el TF-IDF sin lematizar obtuvo mejores resultados

Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo

Además, el clasificador con mejor desempeño fue el SVM linear, por lo que será el usado en los próximos experimentos

--TF-IDF SIN STOPWORDS --

--- SVM (linear) (TF-IDF sin stopwords) ---
Accuracy: 0.6527
              precision    recall  f1-score   support

           0       0.78      0.75      0.77      1863
           1       0.68      0.52      0.59      1070
           2       0.65      0.17      0.28       488
           3       0.54      0.77      0.64      1606

    accuracy                           0.65      5027
   macro avg       0.66      0.55      0.57      5027
weighted avg       0.67      0.65      0.64      5027


-- TF-IDF EN MINUSCULAS --

--- SVM (linear) (TF-IDF en minusculas) ---
Accuracy: 0.6612
              precision    recall  f1-score   support

           0       0.80      0.75      0.77      1863
           1       0.69      0.52      0.59      1070
           2       0.60      0.25      0.35       488
           3       0.55      0.78      0.65      1606

    accuracy                           0.66      5027
   macro avg       0.66      0.57      0.59      5027
weighted avg       0.68      0.66      0.65      5027


-- TF-IDF SIN ACENTOS --

--- SVM (linear) (TF-IDF sin acentos) ---
Accuracy: 0.6616
              precision    recall  f1-score   support

           0       0.80      0.75      0.77      1863
           1       0.69      0.53      0.60      1070
           2       0.60      0.25      0.35       488
           3       0.55      0.78      0.65      1606

    accuracy                           0.66      5027
   macro avg       0.66      0.58      0.59      5027
weighted avg       0.68      0.66      0.65      5027


-- TF-IDF SIN STOPWORDS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin stopwords y en minusculas) ---
Accuracy: 0.6527
              precision    recall  f1-score   support

           0       0.78      0.75      0.77      1863
           1       0.68      0.52      0.59      1070
           2       0.65      0.17      0.28       488
           3       0.54      0.77      0.64      1606

    accuracy                           0.65      5027
   macro avg       0.66      0.55      0.57      5027
weighted avg       0.67      0.65      0.64      5027


-- TF-IDF SIN ACENTOS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin acentos y en minusculas) ---
Accuracy: 0.6616
              precision    recall  f1-score   support

           0       0.80      0.75      0.77      1863
           1       0.69      0.53      0.60      1070
           2       0.60      0.25      0.35       488
           3       0.55      0.78      0.65      1606

    accuracy                           0.66      5027
   macro avg       0.66      0.58      0.59      5027
weighted avg       0.68      0.66      0.65      5027


-- TF-IDF COMPLETO --

--- SVM (linear)  (TF-IDF completo) ---
Accuracy: 0.6533
              precision    recall  f1-score   support

           0       0.78      0.75      0.77      1863
           1       0.68      0.53      0.60      1070
           2       0.66      0.17      0.27       488
           3       0.54      0.77      0.64      1606

    accuracy                           0.65      5027
   macro avg       0.67      0.56      0.57      5027
weighted avg       0.67      0.65      0.64      5027


-- TF-IDF --

--- Logistic Regression (TF-IDF) ---
Accuracy: 0.6563
F1 Score (macro avg): 0.5866
Recall Score (macro avg): 0.5707

--- SVM (linear) (TF-IDF) ---
Accuracy: 0.6370
F1 Score (macro avg): 0.6027
Recall Score (macro avg): 0.6196

--- Decision Tree (TF-IDF) ---
Accuracy: 0.5327
F1 Score (macro avg): 0.4672
Recall Score (macro avg): 0.4639

-- CON LEMATIZAR --

--- Logistic Regression (TF-IDF Lematizado) ---
Accuracy: 0.6606
F1 Score (macro avg): 0.5928
Recall Score (macro avg): 0.5764

--- SVM (linear) (TF-IDF Lematizado) ---
Accuracy: 0.6374
F1 Score (macro avg): 0.6032
Recall Score (macro avg): 0.6205

--- Decision Tree (TF-IDF Lematizado) ---
Accuracy: 0.5443
F1 Score (macro avg): 0.4751
Recall Score (macro avg): 0.4719

El experimento con el TF-IDF con lematizar obtuvo mejores resultados

Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo

Además, el clasificador con mejor desempeño fue el SVM linear, por lo que será el usado en los próximos experimentos

--TF-IDF SIN STOPWORDS --

--- SVM (linear) (TF-IDF sin stopwords) ---
Accuracy: 0.6169
F1 Score (macro avg): 0.5758
Recall Score (macro avg): 0.5873

-- TF-IDF EN MINUSCULAS --

--- SVM (linear) (TF-IDF en minusculas) ---
Accuracy: 0.6374
F1 Score (macro avg): 0.6032
Recall Score (macro avg): 0.6205

-- TF-IDF SIN ACENTOS --

--- SVM (linear) (TF-IDF sin acentos) ---
Accuracy: 0.6348
F1 Score (macro avg): 0.6004
Recall Score (macro avg): 0.6180

-- TF-IDF SIN STOPWORDS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin stopwords y en minusculas) ---
Accuracy: 0.6169
F1 Score (macro avg): 0.5758
Recall Score (macro avg): 0.5873

-- TF-IDF SIN ACENTOS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin acentos y en minusculas) ---
Accuracy: 0.6348
F1 Score (macro avg): 0.6004
Recall Score (macro avg): 0.6180

-- TF-IDF COMPLETO --

--- SVM (linear)  (TF-IDF completo) ---
Accuracy: 0.6155
F1 Score (macro avg): 0.5742
Recall Score (macro avg): 0.5858